### PR TITLE
Add collateral verification for blocks

### PR DESCRIPTION
**Pull request by kriscieslak**

### CODE REVIEW

File: modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/consensus/block/Validator.scala [hunk 1]
```diff
@@ -26,12 +29,18 @@ object Validator {
 
   def isReadyForBlockConsensus(state: NodeState): Boolean = state == Ready
 
+  def withCollateral[F[_]: Applicative, A](collateral: Collateral[F])(peersIds: List[A])(id: A => PeerId) =
+    peersIds.toList.filterA(p => collateral.hasCollateral(id(p)))
```
Comment (marcinwadon) [inline modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/consensus/block/Validator.scala:33]: I think you may achieve `id: A => PeerId` using `Contramap` 🤔

**Review by marcinwadon — APPROVED** [review]

File: modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/consensus/block/Validator.scala [hunk 2]
```diff
@@ -26,12 +29,18 @@ object Validator {
 
   def isReadyForBlockConsensus(state: NodeState): Boolean = state == Ready
 
+  def withCollateral[F[_]: Applicative, A: WithPeerId](collateral: Collateral[F])(peersIds: List[A]): F[List[A]] =
+    peersIds.toList.filterA(p => collateral.hasCollateral(WithPeerId[A].apply(p)))
```
Comment (tbekas) [inline modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/consensus/block/Validator.scala:33]: `toList` is not necessary as `peersIds` is already a list.

File: modules/shared/src/main/scala/org/tessellation/schema/peer.scala [hunk 3]
```diff
@@ -44,6 +44,12 @@ object peer {
       fromId(publicKey.toId)
   }
 
+  type WithPeerId[A] = Function1[A, PeerId]
```
Comment (tbekas) [inline modules/shared/src/main/scala/org/tessellation/schema/peer.scala:47]: It doens't bring much of functionality imo. I would go with simple `_.map(_.id)` function each time when set of peers need to be verified. If you want to be fancy maybe a syntax would be enough, like:
```scala
val peers: List[Peers] = ...
peers.map(_.id).filterCollateralized(collateral)
```

**Review by tbekas — CHANGES_REQUESTED** [review]

File: modules/dag-l1/src/main/scala/org/tessellation/dag/l1/StateChannel.scala [hunk 4]
```diff
@@ -163,7 +166,9 @@ class StateChannel[F[_]: Async: KryoSerializer: SecurityProvider: Random](
           .getTips(appConfig.consensus.tipsCount)
 
         l0Peer <- storages.l0Cluster.getPeers
-          .flatMap(peers => Random[F].shuffleList(peers.toNonEmptyList.toList))
+          .map(_.toNonEmptyList.toList)
+          .flatMap(_.filterA(p => services.collateral.hasCollateral(p.id)))
+          .flatMap(peers => Random[F].shuffleList(peers))
           .map(peers => peers.head)
```
Comment (TheMMaciek) [inline modules/dag-l1/src/main/scala/org/tessellation/dag/l1/StateChannel.scala:172]: Isn't it that right now it may throw and we should handle that?

**Review by tbekas — APPROVED** [review]
LGTM

**Pull request merged**